src/controller/CHIPDeviceController.cpp

```diff
@@ -1809,56 +1822,43 @@ void DeviceCommissioner::OnDeviceConnectedFn(void * context, Messaging::Exchange
-    {
-        // Not the device we are trying to commission.
-        return;
-    }
-
-    if (commissioner->mCommissioningDelegate != nullptr)
-    {
-        CommissioningDelegate::CommissioningReport report;
-        report.Set<OperationalNodeFoundData>(OperationalNodeFoundData(OperationalDeviceProxy(&exchangeMgr, sessionHandle)));
-        commissioner->CommissioningStageComplete(CHIP_NO_ERROR, report);
-    }
+    CommissioningDelegate::CommissioningReport report;
+    report.Set<OperationalNodeFoundData>(OperationalNodeFoundData(OperationalDeviceProxy(&exchangeMgr, sessionHandle)));
+    commissioner->CommissioningStageComplete(CHIP_NO_ERROR, report);
 }
 
 void DeviceCommissioner::OnDeviceConnectionFailureFn(void * context, const ScopedNodeId & peerId, CHIP_ERROR error)
 {
     // CASE session establishment failed.
     DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
+    VerifyOrDie(commissioner->mCommissioningStage == CommissioningStage::kFindOperational);
+    VerifyOrDie(commissioner->mDeviceBeingCommissioned->GetDeviceId() == peerId.GetNodeId());
+    commissioner->CancelCASECallbacks(); // ensure all CASE callbacks are unregistered
 
-    ChipLogProgress(Controller, "Device connection failed. Error %s", ErrorStr(error));
-    VerifyOrReturn(commissioner != nullptr,
-                   ChipLogProgress(Controller, "Device connection failure callback with null context. Ignoring"));
-
-    // Ensure that commissioning stage advancement is done based on seeing an error.
-    if (error == CHIP_NO_ERROR)
+    if (error != CHIP_NO_ERROR)
     {
-        ChipLogError(Controller, "Device connection failed without a valid error code. Making one up.");
-        error = CHIP_ERROR_INTERNAL;
+        ChipLogProgress(Controller, "Device connection failed. Error %" CHIP_ERROR_FORMAT, error.Format());
     }
-
-    if (commissioner->mDeviceBeingCommissioned == nullptr ||
-        commissioner->mDeviceBeingCommissioned->GetDeviceId() != peerId.GetNodeId())
+    else
     {
-        // Not the device we are trying to commission.
-        return;
-    }
+        // Ensure that commissioning stage advancement is done based on seeing an error.
+        // Ensure that commissioning stage advancement is done based on seeing an error.
+        if (error == CHIP_NO_ERROR)
+        {
+            // Ensure that commissioning stage advancement is done based on seeing an error.
+            if (error == CHIP_NO_ERROR)
+            {
+                ChipLogError(Controller, "Device connection failed without a valid error code. Making one up.");
+                error = CHIP_ERROR_INTERNAL;
+            }
 
-    if (commissioner->mCommissioningStage == CommissioningStage::kFindOperational &&
-        commissioner->mCommissioningDelegate != nullptr)
-    {
+            commissioner->CommissioningStageComplete(error);
+            commissioner->CommissioningStageComplete(error);
```

Comment:
Again, why all the extra copies of this?  Shouldn't be here.

Reply:
Yeah this is all part of the same rebase issue.